File: src/main/java/org/sympatico/data/client/db/MongoDocumentLoader.java
package org.sympatico.data.client.db;

import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import org.apache.commons.lang3.tuple.Pair;
import org.sympatico.data.client.db.mongo.MongoDbClient;
import org.sympatico.data.client.db.mongo.MongoRunnable;

import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MongoDocumentLoader {

    private ExecutorService executorService;
    private MongoRunnable[] runnables;
    private static final ConcurrentLinkedQueue<Pair<String, byte[]>> queue = new ConcurrentLinkedQueue<>();

    public MongoDocumentLoader() {
        final MongoDocumentLoader daemonRef = this;
        Runtime.getRuntime().addShutdownHook(new Thread(daemonRef::shutdown));
    }

    public void startMongoDocumentLoader(MongoClientSettings mongoClientSettings, String database, int workerCount) {
        final MongoDbClient mongoDbClient = new MongoDbClient();
        MongoClient mongoClient = mongoDbClient.connect(mongoClientSettings);
        executorService = Executors.newFixedThreadPool(workerCount);
        runnables = new MongoRunnable[workerCount];
        for (int i = 0; i < workerCount; i++) {
            runnables[i] = new MongoRunnable(mongoClient.getDatabase(database), queue);
            executorService.submit(runnables[i]);
        }
    }

    public ConcurrentLinkedQueue<Pair<String, byte[]>> getQueue() {
        return queue;
    }
    private synchronized void shutdown() {
        if (!executorService.isShutdown())
            for (MongoRunnable runnable: runnables) {
                runnable.shutdown();
            }
            executorService.shutdown();
    }
}

File: src/main/kotlin/org/sympatico/data/client/db/mongo/MongoDocumentLoader.kt
package org.sympatico.data.client.db.mongo

import org.slf4j.LoggerFactory
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.time.ExperimentalTime

class MongoDocumentLoader (host: String, port: Int) {

    private val queue: ConcurrentLinkedQueue<Pair<String, ByteArray>> = ConcurrentLinkedQueue<Pair<String, ByteArray>>()
    private var executorService: ExecutorService = Executors.newCachedThreadPool()
    private var runnables: MutableList<MongoRunnable> = mutableListOf()
    private val mongoDbClient = MongoDbClient(host, port)

    fun startMongoDocumentLoader(workerCount: Int, database: String) {
        for (i in 0 until workerCount) {
            runnables.add(MongoRunnable(mongoDbClient.getDatabase(database), queue))
            executorService.submit(runnables[i])
        }
    }

    fun getRunnableQueue(): ConcurrentLinkedQueue <Pair<String, ByteArray>> {
        return queue
    }

    @ExperimentalTime
    @Synchronized
    fun shutdown() {
        LOG.info("shutting down mongo loader...")
        if (!executorService.isShutdown) for (runnable in runnables) {
            runnable.shutdown()
        }
        executorService.shutdown()
        mongoDbClient.shutdown()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(MongoDocumentLoader::class.java)
    }
}